[WIP] - Adicionar integração com o DialogFlow

alexlopes opened the pull request:
**1. O que esse PR faz? Por que precisamos dele?**
Esse PR é responsável por incluir a integração com o DialogFlow, expondo webhooks para receber as request do fullfillment conforme [documentação](https://cloud.google.com/dialogflow/docs/fulfillment-overview). Foi utilizado o [Strategy](https://refactoring.guru/design-patterns/strategy) como padrão de projeto, thanks to @idasilva  

**2. Qual(is) task(s) esta(o) relacionada(s)?**
#11

**3. Como verificar a implementação?**

```sh
go test                
PASS
ok      github.com/gopetbot/messenger   1.530s
```

`go build && ./messenger `

```sh
curl -d '{"originalRequest" : {"source" : "facebook"}  }' http://localhost:8080/v1/pet/webhook -i
HTTP/1.1 200 OK
Date: Tue, 15 Sep 2020 11:46:17 GMT
Content-Length: 34
Content-Type: text/plain; charset=utf-8

{"message":"handled for facebook"}%
```

**4. Informações adicionais para os revisores.**
Estamos trabalhando com a idéia de que o payload que o DialogFlow enviará para nós é nesse [formato](https://stackoverflow.com/a/45073829)

alexlopes committed:
Create HTTP handlers to interact with DialogFlow

- adds handlers
- adds unit tests
- related to #11

idasilva commented on pkg/pet.go:
```diff
@@ -9,5 +11,110 @@ import (
+	Session     string      `json:"session"`
+	ResponseID  string      `json:"responseId"`
+	QueryResult queryResult `json:"queryResult"`
+}
```
@alexlopes, seria legal criar um pacote para o nosso webhook, e consumir através do nosso handler PetProject. oque acha ?

alexlopes replied:
@idasilva acho válido! Boa!

idasilva commented on pkg/pet.go:
```diff
@@ -9,5 +11,110 @@ import (
+}
+
+// getAgentName creates a response for the get-agent-name intent.
+func getAgentName(request webhookRequest) (webhookResponse, error) {
```
@alexlopes eu não acho bacana o prefixo get, poderiamos trocar ? ou manter apenas AgentName?

alexlopes replied:
@idasilva podemos sim, na vdd esse cara provavelmente vai sumir, peguei do exemplo do Google 😅

idasilva commented on pkg/pet.go:
```diff
@@ -9,5 +11,110 @@ import (
+func handleError(w http.ResponseWriter, err error) {
+	w.WriteHeader(http.StatusInternalServerError)
+	fmt.Fprintf(w, "ERROR: %v", err)
+}
```
@alexlopes esse handlererror poderia ser tratado na camada que chama o nosso handler ?

alexlopes replied:
@idasilva onde estaria pensando?

alexlopes committed:
Change way to expose the webhook

- removes pkg directory, changing by middleware package
- adds webhooks using Strategy design pattern
- includes port and host
- adds new lines to gitignore
- fix unit test

idasilva commented on main.go:
```diff
@@ -1,22 +1,22 @@
 
-	mux := handlers.NewRouterInstance()
-	mux.AddRoute(&handlers.Context{
+	mux := middleware.NewRouterInstance("/v1")
```
@alexlopes, gostei do version, não sabia dessa.

idasilva commented on middleware/handler.go:
```diff
@@ -0,0 +1,56 @@
+
+type OriginalRequest struct {
+	Source string `json:"source"`
+}
```
@alexlopes, talvez seja valido colocar esses caras em um arquivo types.

alexlopes replied:
@idasilva concordo, eu até tentei lembrar de um nome e me fugiu o `types.go`

idasilva commented on middleware/router.go:
```diff
@@ -0,0 +1,34 @@
+	Method  string
+	Path    string
+	Handler Handler
+}
```
@alexlopes, poderiamos fazer com que essa struct implement o ServeHTTP, uma função no caso, essa função seria responsavel por delegar atividades aos demais handlers, e se houve algum error poderiamos tratar nela, tirando assim a responsabelidade do "handleResponse"

alexlopes replied:
@idasilva acho que entendi, não lembrava da `ServeHTTP`, vamos ver a melhor forma de fazer então

idasilva commented on middleware/webhook/facebook.go:
```diff
@@ -0,0 +1,119 @@
+
+type message struct {
+	Text text `json:"text"`
+}
```
@alexlopes, talvez colocar em um arquivo types

idasilva reviewed (approved):
@alexlopes,conforme conversamos no whats, podemos fazer o merger assim, e então vou criar uma nova task para ver esses pontos. @moraes1997mateus, @julyanaflores, poderiam validar ?

mmoraes4 reviewed (approved):
🦴 🐕
